erpnext/stock/get_item_details.py

```diff
@@ -396,6 +405,20 @@ def get_item_tax_template(args, item, out):
 			item_group = item_group_doc.parent_item_group
 
 def _get_item_tax_template(args, taxes, out):
+	taxes_with_validity = []
+
+	for tax in taxes:
+		if tax.valid_from:
+			# In purchase Invoice first preference will be given to supplier invoice date
+			# if supplier date is not present then posting date
+			if args.get('bill_date') and tax.valid_from >= getdate(args.get('bill_date')):
+				taxes_with_validity.append(tax)
+			elif not args.get('bill_date') and args.get('posting_date') and tax.valid_from >= getdate(args.get('posting_date')):
+				taxes_with_validity.append(tax)
+
+	if taxes_with_validity:
+		taxes = taxes_with_validity
```

Comment:
`taxes_with_validity` should be ordered by `valid_from` desc and while filtering based on `tax_category` it should take latest one.